### DISCUSSION TITLE
edit Mtext for attrib

### DISCUSSION THREAD

**Discussion by Quasarrs** [mozman/ezdxf#931] [Issues] [answered]
Can not understand! I encountered this problem.
The block contains attributes with multi-line text Mtext.
I'm reading this attribute and trying to rewrite a lot of line text.

` text_data = """
test1
test2
test3
"""

if attrib.has_embedded_mtext_entity:
    editor = MTextEditor().append(text_data)
    m_text = msp.add_mtext(str(editor))
    attrib.embed_mtext(m_text, graphic_properties=True)  
`
Everything works, but if I want to manually change this attribute in the document, then the text changes its position after clicking OK.
How will it lose the attribute insertion point
Maybe I'm adding multiline text to the attribute incorrectly?

**Comment by mozman** [comment]
This code creates a block reference with a multi-line attribute:

```Python
import pathlib
import ezdxf

CWD = pathlib.Path("~/Desktop/Outbox").expanduser()
if not CWD.exists():
    CWD = pathlib.Path(".")


def main():
    doc = ezdxf.new(dxfversion="R2018")
    # create a new block definition:
    block = doc.blocks.new("TEST")
    block.add_circle((0, 0), 1)

    msp = doc.modelspace()
    # create a block reference in modelspace:
    insert = msp.add_blockref("TEST", (0, 0))
    # add a single line attribute to the block reference:
    attrib = insert.add_attrib("ATT0", "dummy")
    # create a MTEXT entity:
    mtext = msp.add_mtext("test1\ntest2\ntest3")
    mtext.set_location((5, 5))
    # embed the MTEXT entity into the attribute:
    attrib.embed_mtext(mtext)

    doc.saveas(CWD / "insert_with_mtext_attribute.dxf")


if __name__ == "__main__":
    main()
```

I can edit the attribute in BricsCAD and the attribute remains at location (5,5):
![image](https://github.com/mozman/ezdxf/assets/4230966/963cf7e4-d320-464e-be6a-d354303b5de8)

**Comment by Quasarrs** [comment]
This, of course, is all great.. But when you already have a created block in the AutoCAD program and you know that it has an attribute with multi-line text that already has a certain formatting. But at the same time you have no idea what its insertion point is.
Your task is simply to transfer the text to this tariff.
At the same time, there is hope for this parameter
graphic_properties – copy graphic properties (color, layer, ...) from source MTEXT if True
And everything seems to work, all formatting is inherited including its position. But when you click on this attribute and edit the test manually and then press ok, then this test from the attribute loses its insertion point and moves out of the drawing.
It feels like it's losing its insertion point. It turns out that it always needs to be transferred? I want it to automatically take the existing coordinate

**Reply by mozman** [reply, marked as answer]
Autodesk introduced many undocumented features like the multi-line ATTRIB in DXF R2018. 

I think Autodesk doesn't really prioritize making different application work well together. So, I'm not too bothered when AutoCAD has problems with certain features that other programs like BricsCAD don't have.

The only Autodesk product I use is DWG TrueView, which uses the AutoCAD engine. This means I can't test AutoCAD thoroughly, especially for issues that only happen in AutoCAD.  I am not willing to pay Autodesk money ever again.

I'm happy as long as DWG TrueView can open DXF files made with ezdxf without crashing. AutoCAD has a very picky and unstable DXF parser, although the DXF documentation states that one should not rely on a fixed tag structure.

This is a multil-line attribute created with BricsCAD:

![image-1](https://github.com/mozman/ezdxf/assets/4230966/1ba6a27d-ff70-4014-b169-649226174f09)

Displayed by TrueView:

![image-2](https://github.com/mozman/ezdxf/assets/4230966/f3f35a3d-850e-47cd-8edc-58b79ce4a1ad)


Even with more resources than I have available, BricsCAD cannot meet the requirements expected by TrueView!

I can add a line to the attribute by ezdxf:

```Python
import ezdxf
from ezdxf.entities import Insert

doc = ezdxf.readfile("ml-attrib.dxf")
msp = doc.modelspace()
insert: Insert = msp.query("INSERT").first

attrib = insert.attribs[0]
if attrib.has_embedded_mtext_entity:
    mtext = attrib.virtual_mtext_entity()
    mtext.text += "\nNewLine4"
    # set only content and location and not graphical properties
    attrib.set_mtext(mtext)

doc.saveas("mtext_attribute_edited.dxf")
```

and BricsCAD shows the expected result:

![image-3](https://github.com/mozman/ezdxf/assets/4230966/4bc381fb-8262-406a-b0f9-17efc8c675ba)

and TrueView of course doesn't:

![image-4](https://github.com/mozman/ezdxf/assets/4230966/5157c0c0-1953-4d9e-af76-55c9aab41cf6)

And there are many other DXF features that have the same problem: MLEADER, ACAD_TABLE, dynamic blocks, ...

In conclusion: 

A round trip process, especially when using features of newer DXF versions such as R2013 and R2018, will not work properly.

Autodesk products often don't work well together with other applications, and they don't really need to fix this because they're the main choice for most people. These issues may continue until more people start using different CAD applications.
